eng/pipelines/release-build-pipeline.yml

```diff
@@ -0,0 +1,249 @@
+parameters:
+  - name: releaseVersion
+    displayName: Version to release, including Microsoft revision suffix (-1) and boring/FIPS suffix (-fips) if they apply.
+    type: string
+
+  - name: releaseIssue
+    displayName: microsoft/go issue number to comment on once complete, or nil.
+    type: string
+    default: 'nil'
+
+  # Allow retrying a build starting at any of these points.
+  - name: poll1MicrosoftGoPRNumber
+    displayName: '1: microsoft/go PR number to poll for merge.'
+    type: string
+    default: nil
+  - name: poll2MicrosoftGoCommitHash
+    displayName: '2: microsoft/go commit hash to poll for build.'
+    type: string
+    default: nil
+  - name: poll3MicrosoftGoBuildID
+    displayName: '3: internal build ID to poll for completion.'
+    type: string
+    default: nil
+  - name: poll4MicrosoftGoImagesPRNumber
+    displayName: '4: microsoft/go-images PR number to poll for merge.'
+    type: string
+    default: nil
+
+  # This parameter intentionally has no default: see release-go-pipeline.yml
+  - name: goReleaseConfigVariableGroup
+    displayName: '[Use "go-release-config" for a real release] Variable group that specifies release target locations and auth.'
+    type: string
+
+trigger: none
+pr: none
+
+variables:
+  - group: go-akams-config
+  - group: go-akams-auth
+  - group: Microsoft-GoLang-bot
+  - group: DotNet-VSTS-Infra-Access
+  # Import config group. This may direct the build to use secrets from the other groups.
+  - group: ${{ parameters.goReleaseConfigVariableGroup }}
+
+jobs:
+  - job: Release
+    # Let the job overall continue 9 hours before failing. The job time has to account for upstream
+    # availability, sync PR validation jobs, mirroring time, and internal build time. Each waiting
+    # step should specify its own timeout that's just long enough for that particular step, so we
+    # don't need to wait 8 hours to get notified about any particular step taking too long.
+    timeoutInMinutes: 540
+    workspace:
+      clean: all
+    pool:
+      # This is a utility job: use generic recent LTS.
+      name: NetCore1ESPool-Svc-Internal
+      demands: ImageOverride -equals Build.Ubuntu.1804.Amd64
+    variables:
+      assetsDir: $(Pipeline.Workspace)/BuildAssets
+      buildAssetJsonFile: $(assetsDir)/assets.json
+      artifactsDir: $(Pipeline.Workspace)/Artifacts
+      microsoftGoInternalPipeline: 958
+      # Set a variable for each polling parameter. If a variable starts off with 'nil', it will be
+      # overridden with a logging command as the pipeline moves on.
+      MicrosoftGoPRNumber: ${{ parameters.poll1MicrosoftGoPRNumber }}
+      MicrosoftGoCommitHash: ${{ parameters.poll2MicrosoftGoCommitHash }}
+      MicrosoftGoBuildID: ${{ parameters.poll3MicrosoftGoBuildID }}
+      MicrosoftGoImagesPRNumber: ${{ parameters.poll4MicrosoftGoImagesPRNumber }}
+    steps:
+      - script: |
+          # Catch and prevent whole-job retries. Retrying the whole job after a
```

Comment:
This is cool. Does AzDO have a notion of 'child' pipelines? Essentially a release would consist of 4 different sub-pipelines, which feed results into the next one?  

I think Jenkins used to have something like that, so you could then retry the individual steps. Not blocking just something to think about.

Reply:
Well, when you have multiple jobs or stages, AzDO does let you retry just the ones that failed. In theory, if all we needed were incremental retries, this could be a single pipeline with individual jobs (or stages) for each retryable chunk. But:

* There's no way to change the go-infra commit run by a retry. (E.g. if a quick pipeline fix is needed before a retry will work.)
* Can't change any variables/parameters. (E.g. if the dev submits their own PR and wants the pipeline to poll that instead of the old, dead PR number.)
  * AzDO UI does let you toggle stages to run/not run when you queue the build, so instead of retrying, you could queue a build, change the parameters, and toggle off the stages that shouldn't re-run, but this is cumbersome.
* Acquiring an agent can take a while. Splitting up jobs multiplies that time.

Some years ago I tried release pipelines and they seemed more flexible for these kind of mid-release changes. But there is no YAML workflow, and they are called "classic" at this point, so best not to add new dependencies on them.

I'll put this in the design doc (in <https://github.com/microsoft/go-infra/pull/43>)--will be useful to have this context around in case AzDO adds new features in the area. 😄